src/GitVersionVsoTask/Run-GitVersion.ps1

```diff
@@ -0,0 +1,66 @@
+param (
+    [string] [Parameter(Mandatory = $false)]
+    $updateAssemblyInfo,
+    [string] [Parameter(Mandatory = $false)]
+    $additionalArguments
+)
+
+Write-Verbose 'Entering Run-GitVersion.ps1'
+
+# Import the Task.Common dll that has all the cmdlets we need for Build
+import-module "Microsoft.TeamFoundation.DistributedTask.Task.Common"
+
+# Returns a path to GitVersion.exe
+function Get-PathToGitVersionExe() { 
+ 	$PSScriptRoot = Split-Path -Parent -Path $MyInvocation.MyCommand.ScriptBlock.File 
+ 	$targetPath = Join-Path -Path $PSScriptRoot -ChildPath "GitVersion.exe"  
+ 	return $targetPath 
+} 
+
+
+## Execution starts here
+
+
+$updateAssms = Convert-String $updateAssemblyInfo Boolean
+
+$loc = $($env:BUILD_SOURCESDIRECTORY)
+$branch = $($env:BUILD_SOURCEBRANCH)
+$commitId = $($env:BUILD_SOURCEVERSION)
+
+$localBranch = $branch.Trim().Replace("refs/heads/", "").Replace("refs/","")
+
+Set-Location $loc
+
+$haveLocalBranch = $false
+# VSO checks out the commit as a detached HEAD
+# loop through and checkout all branches and the one we want
+# then reset merge to get to the desired commit id
+foreach ($remoteBranch in . git branch -r) {
```

Comment:
With the no-fetch option, the state is a detached head as VSO checks out the commit, not the branch. There still needs cleanup. I based my script on the TeamCity GitVersion docs for the case where the clone uses SSH instead of HTTPS.